truffle/src/main/java/org/jruby/truffle/core/exception/NameErrorLayout.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2016 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.jruby.truffle.core.exception;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.object.DynamicObjectFactory;
+import com.oracle.truffle.api.object.dsl.Layout;
+import com.oracle.truffle.api.object.dsl.Nullable;
+import org.jruby.truffle.language.backtrace.Backtrace;
+
+@Layout
+public interface NameErrorLayout extends ExceptionLayout {
+
+    DynamicObjectFactory createNameErrorShape(
+        DynamicObject logicalClass,
+        DynamicObject metaClass);
+
+    DynamicObject createNameError(
+        DynamicObjectFactory factory,
+        Object message,
+        @Nullable Backtrace backtrace,
+        Object name,
+        @Nullable Object receiver);
```

Comment:
Yes I just fixed a problem that was introduced when an exception layout was changed and layout sites weren't updated to match. The error was only apparent in compiled code.

I think you can give exception subclasses different layouts so long as you define your own `#allocate` method for each subclass. We aren't doing that for `SystemCallError` at the moment, but we should be.